File: app/ocr.py
"""Tesseract OCR wrapper with confidence-based filtering."""

import cv2
import numpy as np
import pytesseract
from pytesseract import Output

DEFAULT_LANG = "kor+eng"
MIN_CONFIDENCE = 40


def run_ocr(
    img_bgr: np.ndarray, lang: str = DEFAULT_LANG, min_confidence: int = MIN_CONFIDENCE
) -> tuple[str, list[dict]]:
    """OCR an image and return (full_text, word_list).

    Words below `min_confidence` (Tesseract's 0-100 score; -1 means "not
    text") are dropped. Since handwriting has already been removed from the
    image before this runs, this also acts as a second safety net against
    any inpainting artifacts being misread as text.
    """
    rgb = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB)
    data = pytesseract.image_to_data(rgb, lang=lang, output_type=Output.DICT)

    words = []
    lines: dict[tuple[int, int, int], list[str]] = {}
    for i in range(len(data["text"])):
        text = data["text"][i].strip()
        conf = float(data["conf"][i])
        if not text or conf < min_confidence:
            continue
        key = (data["block_num"][i], data["par_num"][i], data["line_num"][i])
        lines.setdefault(key, []).append(text)
        words.append(
            {
                "text": text,
                "confidence": conf,
                "bbox": [data["left"][i], data["top"][i], data["width"][i], data["height"][i]],
            }
        )

    full_text = "\n".join(" ".join(w) for w in lines.values())
    return full_text, words

File: web/js/ocr.js
/**
 * Tesseract.js wrapper - browser counterpart to app/ocr.py.
 *
 * The worker is created once (language data + wasm core are a multi-MB
 * download) and reused across every page of a PDF.
 */

const DEFAULT_LANG = 'kor+eng';
const MIN_CONFIDENCE = 40;

export async function createOcrWorker(lang = DEFAULT_LANG, onProgress) {
  const { createWorker } = window.Tesseract;
  const worker = await createWorker(lang, 1, {
    logger: onProgress ? (m) => onProgress(m) : undefined,
  });
  return worker;
}

/**
 * OCR a canvas and return (text, words), dropping low-confidence words -
 * mirrors app/ocr.py's run_ocr. Handwriting has already been removed from
 * the canvas before this runs, so this confidence filter is mainly a
 * safety net against inpainting artifacts being misread as text.
 */
export async function runOcr(worker, canvas, minConfidence = MIN_CONFIDENCE) {
  const { data } = await worker.recognize(canvas, {}, { blocks: true });

  const lines = [];
  const words = [];
  for (const block of data.blocks || []) {
    for (const para of block.paragraphs || []) {
      for (const line of para.lines || []) {
        const lineWords = [];
        for (const word of line.words || []) {
          const text = (word.text || '').trim();
          if (!text || word.confidence < minConfidence) continue;
          lineWords.push(text);
          words.push({ text, confidence: word.confidence, bbox: word.bbox });
        }
        if (lineWords.length) lines.push(lineWords.join(' '));
      }
    }
  }

  return { text: lines.join('\n'), words };
}
